faucet/src/faucet.rs

```diff
@@ -202,6 +244,225 @@ pub fn faucet_error(source: CapeWalletError) -> tide::Error {
     })
 }
 
+/// A shared, asynchronous queue of requests.
+///
+/// The queue is a model of an ordered map from public keys requesting assets to the number of
+/// record grants they have received. It is represented as an explicit `HashMap`, which is the
+/// authoritative data structure, as well as an auxiliary, implicit queue in the form of an
+/// unbounded multi-producer, multi-consumer channel.
+///
+/// When a new request comes in, it can be added to the queue with [FaucetQueue::push]. This will
+/// perform validity checks and then add a new entry mapping the public key to 0. It will also send
+/// the public key as a message on the channel. A worker thread will then pick the message off the
+/// channel using [FaucetQueue::pop], and start generating transfers to it. Each time the worker
+/// completes a transfer to the public key, it will call [FaucetQueue::grant], which increments the
+/// counter associated with that public key, persists the change, and instructs the worker to
+/// either continue transferring to the same key or to move on to the next key.
+///
+/// The queue is persistent, so that if the faucet crashes or gets restarted, it doesn't lose the
+/// queue of pending requests. The persistent queue is represented as a log of index entries, of the
+/// form `UserPubKey -> Option<usize>`. An entry `key -> Some(n)` corresponds to updating the
+/// counter associated with `key` to `n`. An entry `key -> None` corresponds to deleting the entry
+/// for `key`. We can recover the in-memory index by simply replaying each log entry and inserting
+/// or deleting into a `HashMap` as indicated.
```

Comment:
It wasn't too hard to understand the code thanks to the explanations here but I wonder if it would make the code more clear by itself if instead of using `Some(0), Some(n > 0), None` to represent the state of a faucet request we had an enum for that.

Reply:
Yep, good suggestion